tools/codegen/gen_functionalization_type.py

```diff
@@ -0,0 +1,210 @@
+        cpp_sig = CppSignatureGroup.from_native_function(f, method=False, fallback_binding=f.manual_cpp_binding)
+        return WARN_TEMPLATE.substitute(
+            op_name=str(f.func.name),
+            cpp_api_name=cpp_sig.most_faithful_signature().name(),
+            maybe_sync_tensor_args=maybe_sync_tensor_args,
+            redispatch_original_args=redispatch_original_args
+        )
+    # call the out-of-place variant of the op
+    functional_cpp_sig = CppSignatureGroup.from_native_function(g.functional, method=False, fallback_binding=f.manual_cpp_binding)
+
+    mutable_input_post_processing = '\n'.join([
+        s for s in [post_process_mutable_input(a) for a in f.func.arguments.flat_non_out] if s is not None])
+    returns_str = return_names_str(f)
+    if len(f.func.returns) > 1:
+        returns_str = f'{dispatcher_sig.returns_type().cpp_type()}({returns_str});'
+
+    return INPLACE_OR_OUT_TEMPLATE.substitute(
+        cpp_api_name=functional_cpp_sig.most_faithful_signature().name(),
+        redispatch_modified_args_remove_out=redispatch_modified_args_remove_out,
+        mutable_input_post_processing=mutable_input_post_processing,
+        returns=returns_str
+    )
+
+METHOD_DEFINITION = CodeTemplate("""\
+${return_type} ${type_wrapper_name}(${formals}) {
+  ${type_definition_body}
+}
+""")
+
+WRAPPER_REGISTRATION = CodeTemplate("""\
+m.impl("${unqual_operator_name_with_overload}",
+       TORCH_FN(${class_type}::${type_wrapper_name})
+);
+""")
+
+# Generates RegisterFunctionalization.cpp
+# These provide the kernels that run the functionalization pass, which can be opted into
+# per backend (e.g. XLA or Vulkan), or as a composable transform (functionalize() in functorch).
+@dataclass(frozen=True)
+class Functionalize:
+    target: Union[
+        Literal[Target.REGISTRATION],
+        Literal[Target.DEFINITION]
+    ]
+
+    @method_with_native_function
+    def __call__(self, g: Union[NativeFunction, NativeFunctionsGroup]) -> List[str]:
+        fs = [g] if isinstance(g, NativeFunction) else g.functions()
+        group: Optional[NativeFunctionsGroup] = None if isinstance(g, NativeFunction) else g
+        outputs = []
+        for f in fs:
+            if get_view_info(f) is None and not modifies_arguments(f):
+                continue
+            if len(f.func.returns) == 0:
+                # TODO: it looks like all the _foreach_ ops fall into this category
+                # Do they.. need to be functionalized?
+                # print("OPERATOR: " + str(f.func.name))
+                continue
+            if get_view_info(f) is not None and modifies_arguments(f):  # view op
+                # TODO: ops that are both views and mutations will require special handling. Pushing off for now.
+                # e.g. transpose_, as_strided_
+                continue
+            if self.target is Target.REGISTRATION:
+                output = WRAPPER_REGISTRATION.substitute(
```

Comment:
Prefer the new style where we use f-strings directly in the function, makes it easier to see what's being generated.